clang/test/Sema/attr-nonblocking-constraints.cpp

```diff
@@ -0,0 +1,256 @@
+// RUN: %clang_cc1 -fsyntax-only -fblocks -fcxx-exceptions -std=c++20 -verify %s
+// These are in a separate file because errors (e.g. incompatible attributes) currently prevent
+// the FXAnalysis pass from running at all.
+
+// This diagnostic is re-enabled and exercised in isolation later in this file.
+#pragma clang diagnostic ignored "-Wperf-constraint-implies-noexcept"
+
+// --- CONSTRAINTS ---
+
+void nb1() [[clang::nonblocking]]
+{
+	int *pInt = new int; // expected-warning {{'nonblocking' function must not allocate or deallocate memory}}
+	delete pInt; // expected-warning {{'nonblocking' function must not allocate or deallocate memory}}
+}
+
+void nb2() [[clang::nonblocking]]
+{
+	static int global; // expected-warning {{'nonblocking' function must not have static locals}}
+}
+
+void nb3() [[clang::nonblocking]]
+{
+	try {
+		throw 42; // expected-warning {{'nonblocking' function must not throw or catch exceptions}}
+	}
+	catch (...) { // expected-warning {{'nonblocking' function must not throw or catch exceptions}}
+	}
+}
+
```

Comment:
Without making any changes specific to these tests, they currently generate:

```
void catches() try {} catch (...) {} // expected-note {{function cannot be inferred 'nonblocking' because it throws or catches exceptions}}

void nb20() [[clang::nonblocking]] {
	catches(); // expected-warning {{'nonblocking' function must not call non-'nonblocking' function 'catches'}}
}

struct S {
    int x;
    S(int x) try : x(x) {} catch (...) {} // expected-note {{function cannot be inferred 'nonblocking' because it throws or catches exceptions}}
    S(double) : x((throw 3, 3)) {} // expected-note {{function cannot be inferred 'nonblocking' because it throws or catches exceptions}}
};

int badi(); // expected-note {{declaration cannot be inferred 'nonblocking' because it has no definition in this translation unit}}

struct A {
    int x = (throw 3, 3); // expected-note {{function cannot be inferred 'nonblocking' because it throws or catches exceptions}}
};

struct B {
    int y = badi(); // expected-note {{function cannot be inferred 'nonblocking' because it calls non-'nonblocking' function 'badi'}}
};

void f() [[clang::nonblocking]] {
    S s1(3);   // expected-warning {{'nonblocking' function must not call non-'nonblocking' function 'S::S'}}
    S s2(3.0); // expected-warning {{'nonblocking' function must not call non-'nonblocking' function 'S::S'}}
    A a;       // expected-warning {{'nonblocking' function must not call non-'nonblocking' function 'A::A'}}
    B b;       // expected-warning {{ 'nonblocking' function must not call non-'nonblocking' function 'B::B'}}
}
```

Reply:
In the last one, the diagnostic currently attaches to:
```
	int x = badi();
```
without the context that one of the constructors is calling this. Hmm...

Reply:
I looked into this a bit more deeply. Member initializers are one variant of a `CXXCtorInitializer`, and currently these are seen as part of the constructor itself, because the AST traversal of the constructor visits their expressions. So violations are attributed to the constructor(s) which use the member initializer, though the source location is correctly obtained from the problem initializer.

It looks like base class initializers and delegating initializers are already being traversed as true function calls. So it's only a normal field or anonymous field that would need special treatment.

If we wanted the language of the diagnostics to distinguish between the body of the constructor and its initializers, one approach would be to treat the initializers as if they were "callable", so that everything about the analysis's diagnostics would see them as "functions" that are "called' from the constructor -- though they'd have to be called "initializers" and not "functions."

That might be too general; a narrower approach would be to note the extra context of being inside an initializer when a violation is generated from a constructor, and add that context to the diagnostic.

Reply:
> a narrower approach would be to note the extra context of being inside an initializer when a violation is generated from a constructor, and add that context to the diagnostic.

I think that sounds more or less like what I had in mind. The main things I was concerned about were
1. We should compltain that an initialiser is e.g. blocking iff the constructor in question actually executes that initialiser, and
2. *Ideally*, it should be obvious to the user—by whatever means—that the problem is in the initaliser.

Reply:
I think the way it is now satisfies those two main concerns:
- the AST traversal of the constructor accurately understands which initializers it invokes
- diagnostics point to the initializer itself as the site of the violation

And yet, the diagnostics could be further improved:
- instead of using "function" everywhere, they could instead use function / blocks / constructor / destructor / lambda / member initializer

I'll take a shot at it.

Reply:
Tracking whether the violation originated from a member initializer, and changing "function" to "member initializer" in the diagnostic is a bunch of busywork but straightforward:

```
warning: 'nonblocking' member initializer must not allocate or deallocate memory [-Wfunction-effects]
   12 |         int *x = new int(42);

note: member initializer cannot be inferred 'nonblocking' because it allocates or deallocates memory
   28 |         int *x = new int(42);
```

In both cases this could be followed by a note "invoked by constructor here"...

Reply:
This turned out to be a bit of a slippery slope. It is relatively straightforward to describe both a call site and a callee as a function / constructor / destructor / member initializer / lambda / block, but it is painful to repeat that `%select{}` construct across a number of diagnostics, and I don't see ways to reduce the number of diagnostics.

I think I'm inclined to keep the change to make the member initializer / function distinction, since that does seem to benefit from the extra "in constructor here" note, but leave aside the specificity about the kind of function-like object.

Reply:
> it is painful to repeat that `%select{}` construct across a number of diagnostics

Er, I’m not 100% sure what you’re referring to. Can you elaborate because maybe I can think of sth?

Reply:
I got this far:

```
def warn_func_effect_violation : Warning<
  "'%0' %select{function|constructor|destructor|constructor's member initializer|lambda|block}1 "
  "must not %select{allocate or deallocate memory|throw or catch exceptions|"
  "have static local variables|use thread-local variables|access ObjC methods or properties}2">,
  InGroup<FunctionEffects>;
def note_func_effect_violation : Note<
  "%select{function|constructor|destructor|member initializer|lambda|block}0 cannot be inferred '%1' because it "
  "%select{allocates or deallocates memory|throws or catches exceptions|"
  "has a static local variable|uses a thread-local variable|"
  "accesses an ObjC method or property}2">;
def warn_func_effect_calls_func_without_effect : Warning<
  "'%0' function must not call non-'%0' function '%1'">,
  InGroup<FunctionEffects>;
def note_func_effect_calls_func_without_effect : Note<
  "function cannot be inferred '%0' because it calls non-'%0' function '%1'">;
```

Notice that the first two have customized "function".  The second and subsequent ones (there's another 6) could do the same thing but they'd have to duplicate those large '%select{}`s. Possibly my aversion to copy/pasting is excessive. But then again I can see other diagnostics that use "function" and could possibly be better if they used the more specific terms too... thus the slippery slope.

Reply:
Hmm yeah, that makes sense. It’s a bit annoying that the wording is just slightly different each time, otherwise, you could use something like `.Summary` to re-use the text of another diagnostic, but I don’t think that works for parts of it, and also, at that point it does really seem like you just want a separate diagnostic.

I think the first two here are fine. We do have a few diagnostics w/ a lot of `%select` cases (e.g. `err_deduced_tst`, `err_expr_not_cce`), but yeah, I’m not sure you can reasonably combine these any further...

Reply:
Also once "member initializer" is carved out for special treatment, there are other possible implicit function calls in the AST that might deserve more precise attribution, e.g. invocation of constructors and destructors for a local variable.

Reply:
> there are other possible implicit function calls in the AST that might deserve more precise attribution

Oh yeah, good idea.

Reply:
There is now special treatment for member initializers, with a note pointing back to the constructor (and special handling if the constructor is implicit). I haven't pursued this idea of more specific attribution of call sites.

Reply:
Sounds good. We can come back to that later (or not if it turns out that the diagnostic is already obvious enough). I think member initialisers (and maybe default arguments too; I forgot what we’re doing for those atm) are the main cases where the location of the code that contains the problem doesn’t make it obvious where it was invoked from.

Reply:
Yeah, we should address that at some point.

Reply:
I poked at this a bit. It looks a little tricky because of the potential for nested `CXXDefaultArgExpr`s, but doable...

Reply:
Also the AST visitor visits the default argument as part of both functions, which I guess is understandable, but for the purposes of this analysis it should only be seen as being evaluated by the caller, right?

Reply:
Implemented. It only gracefully handles one level of `CXXDefaultArgExpr` (no nesting) but that should be a >90% case.

Reply:
> for the purposes of this analysis it should only be seen as being evaluated by the caller, right?

I suppose that’s technically a matter of perspective. *Strictly speaking*, yes, the default argument is only evaluated at the call site—however, an e.g. `nonblocking` function with a `blocking` default argument is probably not a good idea either.

Reply:
This case is insidious!

If the function is declared `nonblocking`, then the default argument's safety should be considered part of the function, because the caller simply trusts the attribute and the function needs to verify it on behalf of all callers.

But when the function is not declared `nonblocking`, and we're trying to infer whether it really is, on behalf of a `nonblocking` caller, then the default argument's safety isn't really the function's problem; it's the caller's problem -- at least in terms of how diagnostics would be phrased, ideally. I'll take another look at this soon.

Reply:
Oops: actually, in the first case, with a `nonblocking` callee, the caller does still examine the `CXXDefaultArgExpr` even if it otherwise trusts the caller's attribute.

Reply:
OK, in the interest of not over-complicating things, I have not tried to detangle the way that default arguments are effectively visited twice, once at the function definition, and again at every call site.

There is however logic so that a warning pointing to a default argument is followed by a note at the call site which invoked it.

Reply:
Yeah, that sounds like it’s enough; we can always revisit this later.